// ng2-adsk-forge-viewer/esm2015/lib/extensions/basic-extension.js
import { fromEvent, merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { Extension } from './extension';
export class BasicExtension extends Extension {
    constructor() {
        super(...arguments);
        this.eventStreams = [];
        this.events = [
            Autodesk.Viewing.FIT_TO_VIEW_EVENT,
            Autodesk.Viewing.FULLSCREEN_MODE_EVENT,
            Autodesk.Viewing.GEOMETRY_LOADED_EVENT,
            Autodesk.Viewing.HIDE_EVENT,
            Autodesk.Viewing.ISOLATE_EVENT,
            Autodesk.Viewing.OBJECT_TREE_CREATED_EVENT,
            Autodesk.Viewing.OBJECT_TREE_UNAVAILABLE_EVENT,
            Autodesk.Viewing.RESET_EVENT,
            Autodesk.Viewing.SELECTION_CHANGED_EVENT,
            Autodesk.Viewing.SHOW_EVENT,
        ];
    }
    static registerExtension(extensionName, callback) {
        BasicExtension.callback = callback;
        super.registerExtension(BasicExtension.extensionName, BasicExtension);
    }
    load() {
        this.events.forEach((eventName) => {
            const obs = fromEvent(this.viewer, eventName).pipe(map(args => this.castArgs(args)));
            this.eventStreams.push(obs);
        });
        this.viewerEvents = merge(...this.eventStreams);
        if (BasicExtension.debugMessages)
            console.log(BasicExtension.extensionName, 'loaded!');
        if (BasicExtension.callback)
            BasicExtension.callback(this);
        return true;
    }
    unload() {
        this.eventStreams = [];
        this.viewerEvents = undefined;
        if (BasicExtension.debugMessages)
            console.log(BasicExtension.extensionName, 'unloaded!');
        return true;
    }
    activate() {
        return true;
    }
    deactivate() {
        return true;
    }
}
BasicExtension.extensionName = 'BasicExtension';
BasicExtension.debugMessages = false;
BasicExtension.callback = null;
