// ngx-nestable/esm5/lib/nestable-expand-collapse-handle/nestable-expand-collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { EXPAND_COLLAPSE } from '../nestable.constant';
var NestableExpandCollapseDirective = /** @class */ (function () {
    function NestableExpandCollapseDirective(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onMouseDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ngxNestableExpandCollapse.item['$$expanded'] = !this.ngxNestableExpandCollapse.item['$$expanded'];
        this._el.nativeElement.dispatchEvent(new CustomEvent(EXPAND_COLLAPSE, {
            bubbles: true,
            detail: this.ngxNestableExpandCollapse
        }));
    };
    NestableExpandCollapseDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxNestableExpandCollapse]'
                },] }
    ];
    /** @nocollapse */
    NestableExpandCollapseDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NestableExpandCollapseDirective.propDecorators = {
        ngxNestableExpandCollapse: [{ type: Input }],
        onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return NestableExpandCollapseDirective;
}());
export { NestableExpandCollapseDirective };
if (false) {
    /** @type {?} */
    NestableExpandCollapseDirective.prototype.ngxNestableExpandCollapse;
    /**
     * @type {?}
     * @private
     */
    NestableExpandCollapseDirective.prototype._el;
}
