// ng2-adsk-forge-viewer/esm2015/lib/component/thumbnail.component.js
import { Component, Input } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { DomSanitizer } from '@angular/platform-browser';
export class ThumbnailComponent {
    constructor(http, sanitizer) {
        this.http = http;
        this.sanitizer = sanitizer;
    }
    ngOnChanges() {
        if (this.thumbnailOptions) {
            this.setImageSrc(this.thumbnailOptions.defaultImageSrc);
            this.getThumbnail(this.thumbnailOptions.documentId);
        }
    }
    getThumbnail(documentId) {
        let url = `https://developer.api.autodesk.com/modelderivative/v2/designdata/${documentId}/thumbnail`;
        let opts = '';
        if (this.thumbnailOptions.width)
            opts += `width=${this.thumbnailOptions.width}&`;
        if (this.thumbnailOptions.height)
            opts += `height=${this.thumbnailOptions.height}&`;
        if (opts)
            url += `?${opts.slice(0, -1)}`;
        this.thumbnailOptions.getAccessToken(this.setAccessToken.bind(this));
        const headers = {
            Authorization: `Bearer ${this.token}`,
            'Content-Type': 'image/png',
        };
        this.http.get(url, { headers, responseType: 'arraybuffer' })
            .subscribe(data => (data) ? this.setImageSrc(`data:image/png;base64,${this.toBase64(data)}`)
            : this.setImageSrc(), error => this.setImageSrc());
    }
    setAccessToken(accessToken, expiryTime) {
        this.token = accessToken;
        this.expire = expiryTime;
    }
    toBase64(data) {
        return btoa(String.fromCharCode(...new Uint8Array(data)));
    }
    setImageSrc(src = '') {
        let imageSrc = '';
        if (src) {
            imageSrc = src;
        }
        else {
            imageSrc = (this.thumbnailOptions && this.thumbnailOptions.defaultImageSrc) || '';
        }
        this.imageSrc = this.sanitizer.bypassSecurityTrustUrl(imageSrc);
    }
}
ThumbnailComponent.decorators = [
    { type: Component, args: [{
                selector: 'adsk-forge-thumbnail',
                template: "<img [src]=\"imageSrc\" />\n"
            },] }
];
/** @nocollapse */
ThumbnailComponent.ctorParameters = () => [
    { type: HttpClient },
    { type: DomSanitizer }
];
ThumbnailComponent.propDecorators = {
    thumbnailOptions: [{ type: Input }]
};
